File: backend/app/controllers/api/v1/contracts.rb
# frozen_string_literal: true

module Api
  module V1
    # S12 / BE-332..BE-334, BE-341..BE-343, BE-347 — **ler e aceitar** contrato.
    #
    # A separação de recursos é a **DEC-38**: aqui vive o recurso `contracts`,
    # que a matriz dá como `R` para os quatro papéis; **publicar** é o recurso
    # novo `contract_versions` (CRUD para OG e Admin), noutro arquivo. No legado
    # não havia recurso nenhum: `contracts_controller.rb` tem 101 linhas e
    # **zero** `before_action`, `may?`, `admin?`, `og?` ou `authorize`, as rotas
    # não têm constraint (`routes.rb:30-31`) e o `create` só carimba o `creator`
    # e salva. **Qualquer autenticado publicava os Termos de Uso.**
    #
    # **DEC-65 — o aceite é AÇÃO, sem bloqueio de acesso.** Nenhum endpoint aqui
    # barra ninguém por pendência; `GET /pending` é o que alimenta o banner
    # persistente e o botão. Ligar o bloqueio numa demo comercial arrisca travar
    # o cliente na primeira tela; o ciclo com bloqueio está registrado para o
    # cutover, com o prazo do jurídico.
    class Contracts < Grape::API
      helpers Api::V1::ControllerHelpers

      helpers do
        def find_contract!
          contrato = ::Contract.find_by(id: params[:id])
          # C1/C-1: id inacessível responde **igual** a id inexistente. Contrato é
          # catálogo global, então aqui não há escopo de projeto — mas a regra de
          # não virar oráculo de existência vale do mesmo jeito.
          error!({ error: 'not_found', message: 'Contrato não encontrado.' }, 404) if contrato.nil?
          contrato
        end
      end

      namespace :contracts do
        before { authenticate_user! }

        desc 'Contratos vigentes, um por tipo' do
          summary 'Contratos vigentes'
          detail 'Uma linha por tipo, com a versão mais recente. O agrupamento vem ANTES da paginação.'
        end
        get '' do
          authorize!('contracts', :read)
          # BE-334: o agrupamento acontece no banco (`DISTINCT ON`), não em Ruby
          # depois do `limit`. Paginar antes de agrupar é o D-20 nesta
          # capability — fazia contrato sumir da lista.
          Api::Entities::Contract.represent(::Contracts::Resolver.current_all)
        end

        desc 'O que este usuário ainda não aceitou' do
          detail 'Alimenta o banner persistente (DEC-65). NÃO bloqueia acesso.'
        end
        get 'pending' do
          authorize!('contracts', :read)
          ::Contracts::PendingService.call(current_user)
        end

        desc 'Uma versão de contrato, com o texto'
        params do
          requires :id, type: String
        end
        get ':id' do
          authorize!('contracts', :read)
          Api::Entities::Contract.represent(find_contract!, type: :full)
        end

        desc 'Aceita a versão vigente do contrato' do
          detail 'Sempre em nome do usuário da sessão. Idempotente. Só a versão vigente. ' \
                 'Isento do gate de somente-leitura (READONLY_EXEMPT_PATHS).'
        end
        params do
          requires :id, type: String
        end
        # PUT, e não POST, porque o efeito é idempotente — e porque é o verbo
        # que o legado já usava (`routes.rb:47`), o que mantém qualquer
        # integração externa funcionando.
        #
        # ⚠️ O caminho `/api/v1/contracts/:id/accept` **tem** de casar
        # `ControllerHelpers::READONLY_EXEMPT_PATHS`. A S0 escreveu o padrão
        # antes de a rota existir; esta é a rota, e o spec
        # `contracts_readonly_spec` prova que casa.
        put ':id/accept' do
          # Sem `authorize!('contracts', :create)`: aceitar não é criar contrato,
          # é exercer um direito do próprio usuário sobre um catálogo global. A
          # matriz dá `R` a todos os papéis, e é `R` que basta — exigir mais
          # trancaria fora exatamente quem precisa aceitar.
          authorize!('contracts', :read)
          contrato = find_contract!

          resultado = ::Contracts::AcceptService.call(
            user: current_user, contract: contrato,
            ip_address: request.ip, user_agent: request.user_agent
          )
          if resultado.status >= 400
            error!({ error: resultado.error, message: resultado.message, code: resultado.code }.compact,
                   resultado.status)
          end

          status resultado.status
          Api::Entities::ContractDeal.represent(resultado.deal)
        end
      end

      # ---------------------------------------------------------------------
      # `/api/v1/me/terms` — o aceite em bloco, do próprio usuário.
      # ---------------------------------------------------------------------
      # É a segunda rota que `READONLY_EXEMPT_PATHS` isenta, e ela existe por um
      # motivo concreto: o banner da DEC-65 pede o aceite de **tudo o que está
      # pendente** num clique. Sem isto, um usuário com dois contratos pendentes
      # teria de aceitar duas vezes, e o banner ficaria de pé depois do primeiro
      # clique parecendo que não funcionou.
      namespace :me do
        before { authenticate_user! }

        namespace :terms do
          desc 'Situação dos Termos para o usuário da sessão' do
            detail 'Pendências + histórico de aceites, inclusive os `implicit_legacy` (DEC-66).'
          end
          get '' do
            authorize!('my_account', :read)
            {
              pending: ::Contracts::PendingService.call(current_user),
              accepted: Api::Entities::ContractDeal.represent(
                ::Contracts::PendingService.history(current_user)
              ).as_json
            }
          end

          desc 'Aceita de uma vez tudo o que está pendente'
          post '' do
            authorize!('my_account', :read)
            resultado = ::Contracts::AcceptService.accept_all_pending(
              user: current_user, ip_address: request.ip, user_agent: request.user_agent
            )
            if resultado.status >= 400
              error!({ error: resultado.error, message: resultado.message, code: resultado.code }.compact,
                     resultado.status)
            end

            status 200
            {
              pending: ::Contracts::PendingService.call(current_user),
              accepted: Api::Entities::ContractDeal.represent(
                ::Contracts::PendingService.history(current_user)
              ).as_json
            }
          end
        end
      end
    end
  end
end

File: frontend/src/lib/api/contracts.ts
import { apiClient } from './client'
import { pageParams, readPageMeta, type PageMeta, type PageRequest } from './pagination'

/**
 * Contratos — Termos de Uso e Política de Privacidade.
 *
 * **Dois recursos, de propósito (DEC-38):**
 *
 *  - `contracts` — **ler e aceitar**. `R` para os quatro papéis.
 *  - `contract_versions` — **publicar e administrar**. CRUD só para OG e Admin.
 *
 * O legado não tinha recurso nenhum: `contracts_controller.rb` tem 101 linhas e
 * zero autorização, e as rotas não têm constraint. Qualquer autenticado
 * publicava os Termos de Uso. **O gate é do servidor** — esta camada só
 * organiza as chamadas; esconder o botão não autoriza nada.
 *
 * **DEC-65 — o aceite é AÇÃO, sem bloqueio.** Nada aqui interrompe navegação:
 * `pending()` alimenta o banner persistente e o botão. Bloquear o acesso numa
 * demo comercial trava o cliente na primeira tela.
 */

export type ContractKind = string

export interface Contract {
  id: string
  kind: ContractKind
  slug: string
  title: string
  version: number
  published_at: string
  created_at: string
  updated_at: string
  /** Só no detalhe. Já vem **sanitizado** pelo servidor (allowlist do BE-345). */
  description_html?: string
  creator?: { id: string; name: string } | null
  /** Só para quem administra. */
  accepted_count?: number
  divergent_count?: number
}

export interface PendingContract {
  id: string
  kind: ContractKind
  slug: string
  title: string
  version: number
  published_at: string
  /** Informativo: passado o prazo a interface insiste mais. **Não bloqueia.** */
  tolerance_until: string
  overdue: boolean
}

/** DEC-66 — o histórico distingue quem aceitou de quem foi carimbado. */
export type DealSource = 'explicit' | 'implicit_legacy'

export interface ContractDeal {
  id: string
  contract_id: string
  kind: ContractKind
  version: number
  accepted_at: string
  source: DealSource
  legacy_accepted_at: string | null
  hash_matches_current: boolean
}

export interface TermsStatus {
  pending: PendingContract[]
  accepted: ContractDeal[]
}

export interface ContractCatalogEntry {
  kind: ContractKind
  slug: string
  current_version: number | null
  versions_count: number
  accepted_count: number
}

export interface ContractPrefill {
  kind: ContractKind
  slug: string
  next_version: number
  title: string
  description_html: string
  previous_version: number | null
  previous_id: string | null
  current_accepted_count: number
}

export interface ContractImpact {
  accepted_count: number
  divergent_count: number
  is_current: boolean
}

export interface ContractVersionsPage {
  versions: Contract[]
  meta: PageMeta
}

export interface ContractVersionFilters extends PageRequest {
  kind?: string
  q?: string
}

/** Leitura e aceite — o que qualquer papel alcança. */
export const contractsApi = {
  /** Uma linha por tipo, com a versão vigente. */
  list: () => apiClient.get<Contract[]>('/api/v1/contracts'),

  get: (id: string) => apiClient.get<Contract>(`/api/v1/contracts/${id}`),

  /** O que alimenta o banner. Nunca bloqueia navegação (DEC-65). */
  pending: () => apiClient.get<PendingContract[]>('/api/v1/contracts/pending'),

  /**
   * Aceita UMA versão. `PUT` porque o efeito é idempotente — e porque é o verbo
   * que o legado já usava (`routes.rb:47`).
   *
   * O usuário **não viaja no payload**: é sempre o da sessão. No legado
   * `user_id` estava no `permit` e o primeiro `create` podia gravar aceite em
   * nome de outro (D-68).
   */
  accept: (id: string) => apiClient.put<ContractDeal>(`/api/v1/contracts/${id}/accept`),

  /** Situação dos Termos para o usuário da sessão: pendências + histórico. */
  myTerms: () => apiClient.get<TermsStatus>('/api/v1/me/terms'),

  /** Aceita TUDO o que está pendente num clique — é o botão do banner. */
  acceptAllPending: () => apiClient.post<TermsStatus>('/api/v1/me/terms'),
}

/** Publicação e administração — OG e Admin (DEC-38). */
export const contractVersionsApi = {
  list: async (filters: ContractVersionFilters = {}): Promise<ContractVersionsPage> => {
    const params: Record<string, string | number> = { ...pageParams(filters) }
    if (filters.kind) params.kind = filters.kind
    if (filters.q) params.q = filters.q

    const resposta = await apiClient.getRaw<Contract[]>('/api/v1/contract_versions', { params })
    return {
      versions: resposta.data ?? [],
      meta: readPageMeta({ body: resposta.data, headers: resposta.headers as any }, filters),
    }
  },

  catalog: () => apiClient.get<ContractCatalogEntry[]>('/api/v1/contract_versions/catalog'),

  get: (id: string) => apiClient.get<Contract>(`/api/v1/contract_versions/${id}`),

  /** O primeiro contrato de um tipo abre VAZIO — no legado esse caso estourava. */
  prefill: (kind: string) =>
    apiClient.get<ContractPrefill>('/api/v1/contract_versions/prefill', { params: { kind } }),

  /** Mitigação 2 da DEC-80: quantos aceites ficam com hash divergente. */
  impact: (id: string) => apiClient.get<ContractImpact>(`/api/v1/contract_versions/${id}/impact`),

  create: (payload: { kind: string; title: string; description: string }) =>
    apiClient.post<Contract>('/api/v1/contract_versions', payload),

  update: (id: string, payload: { title?: string; description?: string }) =>
    apiClient.put<Contract>(`/api/v1/contract_versions/${id}`, payload),

  remove: (id: string) => apiClient.delete<{ success: boolean }>(`/api/v1/contract_versions/${id}`),

  /**
   * Baixa a prova de aceite (OPS-333).
   *
   * **Não é `<a href>`.** O access token vive em memória, não em cookie: uma
   * navegação direta ao endpoint sai sem `Authorization` e volta 401 — e o
   * usuário veria uma tela de erro em vez do arquivo. O download passa pelo
   * `apiClient` (que anexa o token) e vira `blob` local.
   */
  downloadProof: async (id: string, filename: string) => {
    const blob = await apiClient.get<Blob>(`/api/v1/contract_versions/${id}/proof`, {
      responseType: 'blob',
    })
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    link.setAttribute('download', filename)
    document.body.appendChild(link)
    link.click()
    link.parentNode?.removeChild(link)
    window.URL.revokeObjectURL(url)
  },
}

export interface PublicContract extends Contract {
  /** Caminho resolvido pelo SERVIDOR, a partir da allowlist. Nunca o parâmetro. */
  return_to: string
  return_to_allowed: boolean
}

/**
 * Superfície pública — leitura sem sessão.
 *
 * **D-69**: `returnTo` é uma **chave** (`login`, `console`, `profile`, `faq`),
 * nunca uma URL. O legado interpolava o parâmetro na view e o usava como
 * destino: XSS refletido **e** open redirect na mesma variável. Aqui o cliente
 * manda a chave e recebe de volta o caminho — texto do visitante não vira href
 * em lugar nenhum.
 */
export const publicContractsApi = {
  kinds: () =>
    apiClient.getPublic<{ kinds: { kind: string; slug: string }[]; return_destinations: string[] }>(
      '/api/v1/public/contracts',
    ),

  get: (kindOrSlug: string, returnTo?: string) =>
    apiClient.getPublic<PublicContract>(
      `/api/v1/public/contracts/${encodeURIComponent(kindOrSlug)}`,
      returnTo ? { params: { return_to: returnTo } } : undefined,
    ),
}
